### DISCUSSION TITLE
Roadmap to V1

### DISCUSSION THREAD

**Discussion by mscolnick** [marimo-team/marimo#2899] [Announcements]
We're building a unified toolchain for working with data. On the road to marimo version 1.0, we'll focus on enabling the following goals:

1. **Rapid data manipulation** with Python and SQL
2. Excellent **developer experience in the marimo editor**
3. Excellent **developer experience** **working with notebook files**
4. Seamless **embedding** of notebooks in web pages and **sharing** notebooks as applications
5. Working with data and compute **at any scale**
6. **Enterprise-readiness**

Here's a breakdown of our roadmap, organized into these categories.

## Rapid data manipulation

- Reactively respond to non-python file changes (e.g. parquet, csv, json) #3258 
- More comprehensive parsing of the SQL dataflow graph #3103
- Support multiple SQL dialects #3201
- Easily connect to or import data from databases, open tables, and blob storage #3737 
- Interactive domain-specific elements for machine learning and data science, such as an array viewer
- High-performance interactive dataframe viewer - dataframe viewer's column chart are aggregated on the backend #3104
- High-performance datasource panel - datasource panel's column chart is aggregated on the backend #3105

## Developer experience in the marimo editor

We'll be working on improving our **code intelligence**, including more robust code completion and the addition of type-checking and linting; on adding **real-time collaboration** to the editor; building out more AI assistance features; and improving usability across the board.

### Code Intelligence

- Language Server Protocol support
    - https://github.com/marimo-team/marimo/issues/1163
    - https://github.com/marimo-team/marimo/issues/1963
- Improved go-to-definition
    - https://github.com/marimo-team/marimo/issues/1430
    - #3857
- Variable renaming #3856
- Bring your own LSP servers: ruff, pyright

### Multiplayer

- Real-time collaboration 
- Local network sharing #3106

### AI integration

- Chat sidebar panel #3084
- AI inline edits #3085
- Custom models for AI autocompletion #2193 
- RAG marimo docs and common libraries #4151 

### General Usability

- Improved error messages in the UI #3110
- Improved error messages in the terminal #3111
- Auto-fix for common errors #3343
- Migrate to react-aria-components for better ARIA/accessibility.
- Multi-column notebooks #3112 #2531
- Helper panel for previous cell runs (times, success/failure) #2898
- Easier installation of optional dependencies that enable SQL support (duckdb), AI features (openai/anthropic), Charting (altair), Charting performance (vega-fusion) #3113
- Improved UX when auto-instantiate (no autorun on startup)
    - #3083
    - #2885
    - #2983
    - #3082

### Working with the marimo notebook file

- Direct editing of notebook files in your text editor of choice, with side-by-side rendering of notebook outputs #3114 #2675 #1511
- Top-level functions in notebook files #2293
- Making the notebook file work better with type checkers and linters
    - #2325
- Making it possible to declare arguments, with autogenerated CLI help, for notebooks that are run as scripts https://github.com/marimo-team/marimo/pull/4306
- VSCode plugin

### Package management

- Full implementation of PEP 723-style metadata, tightly integrated with the uv package manager #3115 #3827

### Configuration

- pyproject.toml support for `[tool.marimo]` #3069
- App-level configuration overrides / partial configuration #2885

### Sharing notebooks on GitHub

- GitHub viewer
    - #2960
    - #2961
- GitHub pages
    - `marimo export html-wasm --mode=read|edit` #3116

### Static-site generation

- Export to ipynb for compatibility with existing tools #3048 #3133
- Export to markdown with `--include-outputs`

### Building and deploying data apps from marimo notebooks

- Improvements to deploying marimo notebooks as apps via FastAPI
    - #2938
    - #3095
    - #3619

### Execution

- Remote execution (bring-your-own-backend)
- Improve WebAssembly cold starts
- [exploratory] Async/concurrent execution
- [exploratory] Parallel execution #1103
- [exploratory] Distributed (”hybrid”) execution
- [exploratory] Parametrized/batched execution

### Caching

- Remote caching for `mo.persistent_cache()`
    - S3 backend
    - Redis
- Caching robustness
- Object serialization options (pickle, parquet)
- Cache storage format (file, SQLite)
- [exploratory] Resuming notebooks after they have been shutdown
    - Opt-in cell-level caching

### Memory Consumption

- Lazy construction of various objects

### Tracing

- Execution tracing/telemetry and identifying slow cells #2898

### Pipelines

- Proof-of-concept implementation using marimo notebooks in pipelines
- Integration with Airflow

### Stability

- Internal: better visual snapshot testing for visual regressions
- Internal: more comprehensive testing for performance regressions

### Security

- Secret management
- Notebook trust model

### Instrumentation

- Custom OTEL instrumentation hook
- Custom logging
- Metrics collection/endpoint
- Custom <head> scripts (e.g analytics) #2926

**Comment by raspharaoh** [comment]
- [x] Quarto, Zotero plugin or other reference manager.

Better support for references in aid of reproducibility. There is a feature in Marimo where typing a colon ':' produces an icon/emoji list to choose from. It would be good if there was a more general way or another hotkey to produce a popup custom list from a file of items.

Best wishes for 1.0.

**Reply by dmadisetti** [reply]
Pretty back logged, but I did make an issue to track this: https://github.com/marimo-team/marimo/issues/4756

Quarto is available!

**Comment by Haleshot** [comment]
> ## Community
> ### Integrations

Would be nice to have a [hugoblox](https://hugoblox.com/) plugin; although I wonder if you can do something readily now (as it works well with markdown; renders mermaid diagrams well in blog posts).

> ### Shareability
> * Have a recommended way to write blog posts powered by marimo easily

I'd really like to shoutout Weights and Biases's (wandb) way of writing [reports/blogs](https://wandb.ai/site/reports) where you can [see](https://wandb.ai/stacey/deep-drive/reports/The-View-from-the-Driver-s-Seat--Vmlldzo1MTg5NQ?_gl=1*exlvtl*_ga*MjkwNTU3MDA3LjE3MTgzNzQzMDE.*_ga_JH1SJHJQXJ*MTczMjM0NzcyOC4zLjEuMTczMjM0Nzc1NS4zMy4wLjA.*_ga_GMYDGNGKDT*MTczMjM0NzcyOC4zLjEuMTczMjM0Nzc1NS4wLjAuMA..&utm_source=fully_connected&utm_medium=blog&utm_campaign=view%20from%20the%20drivers%20seat) how they display their plots (interactive); good example of dogfooding?

**Comment by alexscordellis** [comment]
First of all I want to say I'm very excited about marimo. It's giving me very strong "polars circa late 2021" vibes: clearly a better solution shape to the problem than the incumbent, but a few rough edges mean it's not quite ready for primetime yet. Hopefully some of these get ironed out and I look forward to watching it spread across teams in my company the way that polars has over the last couple of years.

From trying marimo as a replacement for JupyterLab over the last few weeks, here are some improvements I would love to see and which would help with adoption.

1. **Rendering of large dataframes** Rendering large dataframes often crashes my browser. I think this is because marimo currently renders dataframes to JSON on the server then sends this to the browser. I often work with dataframes with millions or tens of millions of rows, to which this approach doesn't scale. Of course I can call `.head()` before rendering, but this is a new habit to learn and currently the penalty for forgetting is a hard reload of all marimo tabs, which is just too large. The dataframe rendering is a cut above that in Jupyter: the little histograms (especially the null counts) are super useful, as is the ability to page through a small dataframe in the browser - this beats hunting for the relevant rows with `.filter()`. So I'd hope server side paging can retain all those features while handling large dataframes better.
2. **Ability to clear cell outputs independent of computation state** I often have many interactive plotly charts rendered in a notebook session. At some point this becomes too much for the browser and plotly starts misbehaving. The solution in JupyterLab is to clear the outputs you aren't looking at any more. In marimo, the only way I can find to do this is to comment out the contents of the relevant cell and re-run it, which isn't good for reproducibility, and can break the dependency graph if those cells exported variables. I'd like marimo to let me (by mouse or keyboard shortcut) clear the display of the cell output (either individually or for entire notebook) _without changing the state of the variables the cell(s) defined_.

These first two items above are blocking me from moving to marimo as my default notebook engine. The issues here are likely common to many of my colleagues so I think addressing these would really enable wider adoption:

Further suggestions:

1. **Run all prerequisites of this cell**. If I open a notebook I've not used in a while, often it's to look at one particular chart or other output. Right now I have to work out for myself which cells I need to run first to get to the one I'm after. It would be great if I could just find the relevant cell and click "run all predecessors and this", and it would run the minimal set of predecessor cells. This feels like a great way to make the most of marimo's execution model.
2. **Cell runtime history**. Something I miss from JL (specifically from https://pypi.org/project/jupyterlab-execute-time/) is being able to see how long a cell took on previous executions. It's really useful to know how long to expect a cell to take before triggering execution. (I can also think of cool ways to combine this with the execution graph e.g. "to get to this cell from the initial state takes X minutes").
3. **Display the time at which cell was last run** Useful if you've altered some external state e.g. data files on disk, and you want to know which of your cells have been run since.
4. **Configure copilot to go via a proxy** I work on machines without a direct internet connection, but have access to proxies to allow copilot to connect from e.g. vscode, JetBrains IDEs. Would be great if I could do the same from marimo.


I hope that's helpful, and look forward to seeing marimo grow and develop.

**Reply by alex-scordellis-izone** [reply]
Happy to discuss / provide more info / ticket up any of the above if that would be useful

**Reply by mscolnick** [reply]
Hi Alex, thank you for all the feedback. All of this is on our roadmap and should be done early one. I will make sure they are captured in the above outline. Going to provide some small responses to each:

**1. Rendering of large dataframes**. We do currently paginate the table, so it should only be sending `.head()` to the frontend for the rows. However, for the column summaries, we send `20_000`, which might be a cause for your slowness. It is on the roadmap to fix the performance (aggregate on the backend). In the short term, you can disable this with `show_column_summaries=False` in `mo.ui.table`, or you can change the **User Settings > Display > Dataframes > Rich -> Plain**

**2. Ability to clear cell outputs independent of computation state**
This is a great idea, I can work on this today as long I as don't hit any hiccups.

**3. Run all prerequisites of this cell.** We plan on improving "un-initialized notebook" when `on-startup='lazy'`. Maybe we can incorporate is there. In the short-term, I will add a right-click to the `Run Button` to run ancestors. We often can figure this out from the error type ("your referenced a variable that existed in code, but has not yet been run", i can surface this special run button there)

**4. Cell runtime history/Display the time at which cell was last run**. We plan to add a whole sidebar helper for this. In the short-term, I will add a hover-tooltip to the existing timer which can show previous runs and the last time.

**5. Configure copilot to go via a proxy**. On the roadmap, I will reach out to you when this is done. Hopefully you can help test this for us. 


I will provide some short responses to each.

**Reply by alexscordellis** [reply]
Amazing, thanks for the quick feedback. I'll try out the dataframe tips. Happy to test out the new features as they come along.

**Reply by akshayka** [reply]
@alex-scordellis-izone , any chance you have a reproduction of the performance issue for rendering large dataframes? If so that would really help. I have seen this occasionally but haven't been able to reproduce it recently.

Thanks for all the feedback! It's very helpful.

**Reply by alex-scordellis-izone** [reply]
All the specific datasets I work with are proprietary and internal, but I can try to capture any common characteristics that seem to cause this and let you know what I find

**Reply by akshayka** [reply]
I can reproduce an issue with rendering dataframes with many columns (~10k), but the number of rows has no effect on performance in my reproduction. @alex-scordellis-izone, I'd be curious if your performance issues are also with wide dataframes, or if you've hit performance issues with narrow dataframes as well (~100 columns or fewer).

EDIT: Reproduction is here - https://github.com/marimo-team/marimo/issues/2991

**Reply by AdrienDart** [reply]
Hi, I also get the issue with rendering large dataframes that crash the browser tab (sometimes with less than 50 columns). I don't have any reproduction unfortunately.

**Reply by mscolnick** [reply]
@AdrienDart - what version are you on? the latest version should have a fair amount of improvements for this. 

@alex-scordellis-izone, can you also try the latest marimo version (0.9.29) with your large dataframes.

**Reply by akshayka** [reply]
@alex-scordellis-izone, have you gotten a chance to try the latest version of marimo with your large dataframes? We're very curious to hear whether you are still hitting performance issues.

**Reply by AdrienDart** [reply]
Hi, sorry for the late reply. I can say that browser crashes have been less frequent but do sometimes happen with a lot of dataframes displayed (sometimes along altair charts) and a few Marimo notebooks open. Quick question for you, when I want the uncropped version of a table, should I use . pipe(mo.ui.table, max_columns=None) or is there an easier way to achieve this? Shall we get at some point a toggle near each table to change page size (as pycharm gives us the option, see (https://www.jetbrains.com/help/pycharm/settings-tools-database-data-views.html#limitations)) or number of columns displayed?

**Comment by michaelsch71** [comment]
Wow, looks like a very ambitious and promising roadmap. I already love marimo and I am really looking forward to what the future will bring. Since you have mentioned the plugin for VSCode I personally would like to see a plugin for PyCharm as well, which I feel like is one of the goto IDE's for python developers.

I'm looking forward to see what the future of marimo will bring. Thanks for all your passion for this amazing project!

**Reply by m-aciek** [reply]
For what it's worth I've opened a feature request for a PyCharm plugin, to track the demand: https://github.com/marimo-team/marimo/issues/6297

**Comment by solarslurpi** [comment]
A [model context protocol](https://modelcontextprotocol.io/introduction) client.  How amazing this will be to be able to plug in our data sources and use AI on them.  The client should be able to be configured not just with paid LLMs like Claude and OpenAI, but Ollama so that easy integration with fairly powerful locally run LLMs are supported.

Marimo is very exciting. Thank you for a brief moment of sunshine.

**Reply by Haleshot** [reply]
An update on this:
_"CodeMirror extension that implements the Model Context Protocol (MCP) for resource mentions and prompt commands."_ -  https://github.com/marimo-team/codemirror-mcp

**Comment by AdrienDart** [comment]
Hi, 

Congratulation on your latest funding! Well deserved! 

I have a few suggestions for you on top of the already mentionned ones. Apologies in advance if some of the features are already implemented.

1. A toggle button on each cell to trigger a plain mode (to avoid resorting to mo.plain manually)
2. The ability to copy to the clipboard lists, dataframes, columns of dataframe, plots
3. %%timeit magic
4. Multiple cells selection/operation (to then delete, move, copy, cut, paste, etc.)
5. Some debugging features
6. Set the default number of lines displayed in the dataframe as a setting
7. Ability to add project dependencies (I currently add sys.path.insert(0,path_to_other_project) in the first cell of my notebook)
8. A terminal?
9. Tabs in a jupyterlab fashion

Thank you!

**Comment by schlich** [comment]
Piggybacking on number 8 above!

I know terminal-things are not on the v1 roadmap for what are likely very good and pragmatic reasons, and I wonder if this isn't the right space for "wish list" requests (if there is one, i'll be happy to move it there...)

... but.  i think i MIGHT be able to die happy if marimo made a terminal emulator (Ha, ok, fine, fine, i'll settle for a TUI). If y'all can do anywhere close the same thing in the terminal as y'all did with the notebook experience, it'd be an incredible win.

Excel watch out? 🤤

Imagine the best of all worlds with me:
- The reactivity and ergonomics of Marimo
- The performance and stability of a modern Terminal e.g. Kitty/Wezterm/Ghostty
- Window Management and context switching e.g. Zellij
- Dataframe support and pipeline scripting w/ Nushell?
- Hotkey ergonomics of Helix/Kakoune (motions first!)
- Simplicity of Harlequin for a traditional catalog

... all on one package?
Sure would pay a hefty premium for that 😉

Regardless, lots of groundwork already laid!  Seems like too many moving parts to manage via various plugin systems. Integrated is cooler anyways.  

Thanks to the Marimo team for this amazing tool which has already had a seismic impact on my productivity. Very excited to see what the future holds.

**Comment by codemee** [comment]
I'd like to have a side by side layout that show just one code cell on left and the output on the right with pageup/down or other UI to nivigate prevous/next cell. This would be helpful for instructor to demo or teach coding. Thanks.

**Reply by mscolnick** [reply]
I think we could extend the Slides layout to do this, such that we can show the code of the cell that is being presented

**Reply by codemee** [reply]
That's cool. Wish that would happen soon.

**Comment by fgeorgepar** [comment]
**Developer experience in the marimo editor:** It would be useful from a developer experience and analytics standpoint if the mo.ui.dataframe() transformations and any other UI/widget changes made during exploratory analyses were persisted after saves of the notebook. Alternatively, allow users to create some sort of configuration file that captures changes to these elements that can be manually saved and reloaded by the user.  

It would significantly improve the exploratory features of marimo if the full state of the notebook was saved in this way.

**Reply by mscolnick** [reply]
Hi @fgeorgepar - have you tried [`mo.persistent_cache`](https://docs.marimo.io/guides/expensive_notebooks/?h=cache#save-and-load-from-disk-with-mopersistent_cache)?

**Reply by fgeorgepar** [reply]
Hey @mscolnick - I used the data transformation UI to select some columns using the code below, but it throws an error: marimo._save.cache.CacheException: Failure during save.

```
import marimo as mo
with mo.persistent_cache(name="my_cache"):
    df = mo.ui.dataframe(embed_df)
df
```

**Reply by fgeorgepar** [reply]
Also: An exception was raised when attempting to cache this code block with the following message:
Can't pickle local object 'PandasTableManagerFactory.create.<locals>.PandasTableManager'
NOTE: The cell has run, but cache has not been saved

**Reply by mscolnick** [reply]
You won't want to cache the UI element, but likely the ".value" of the element.

**Reply by fgeorgepar** [reply]
Ah, thanks. I gave that a try but there were three problems: (1) trying to create the dataframe and cache it using the .value in the same cell throws an error, (2) creating two cells like the following started some sort of ongoing function call that blocks the notebook from running.
cell1
```
filtered_data = mo.ui.dataframe(embed_df)
filtered_data
```

cell2
```
with mo.persistent_cache(name="my_cache"):    
    filtered_data.value
```

(3) After restarting the notebook and editor, I noticed I could get it to run, but it still did not store the dataframe results in either the UI element or by accessing the value downstream.  I'm not sure it would necessarily be a usable workflow if it worked since now when you repopen the notebook you'd have two cells, the second with your filtered dataframe and the first cell blank without any of the UI selections saved. It would probably still lead to messy notebooks and anyone receiving the notebook might be confused as to why those cells are empty.  I'm not sure if a solution like hex.tech is possible where the selections themselves are cached without any workarounds.  

Let me know if I'm missing something and this can be done easily in a different way. Thanks.

**Reply by dmadisetti** [reply]
As is, this won't work because cache expects to compute values, not just hold on to them. You can click the `python code` tab, and paste that result under cache: https://marimo.app/l/2z31d7

But it might be easy to add a: `mo.ui.dataframe(df, cached=True)`, which could just do this for you under the hood. Same could be true for the experimental `mo.ui.data_editor`

**Comment by trenchant-dev** [comment]
I'm similar to george above. I think missing from the roadmap is "Find two or three devs who've never used marimo, watch them learn to use it while giving 0 input". That's a game that yields important gains for more iterations than you'd expect. Here is a collection of little things that I've noticed.

1. As george pointed out, the checkboxable ui.table thing is nowhere near as persistent as I expected. I think what I should have done was `mo.ui.dataframe` and then copied the python code when I was done tinkering.
2. In some table views I can plaintext search across all rows with a little search bar at the bottom, in others not. 
3. Similarly, in some I can click a column and then Filter to filter within that column, in others not.
4. I was trying to use the new `format_mapping` feature to cash-format my amounts, but this resulted in "NaNs" shown instead (any string format that yields something non-float-parseable does this, apparently).
5. The per-notebook autoRun will be big. I tried to use Marimo at work for tinkering with side-effect-causing commands in an ops notebook, but it was simply too unpredictable when a cell was going to be run (maybe at startup, maybe reactively, etc). I found it inconvenient to wrap everything in buttons.
6. Not being able to `display` multiple things from one cell is rather odd. I eventually found that `vbox` might be a neat solution and went to chip in [here](https://community.marimo.io/display-output-of-all-expressions-in-a-single-cell-qPyzBXj63kyW), but can't get Discord to cooperate with my account.
7. If only going to autodisplay one thing per cell, there should probably be a handy "split cell" in the right-click menu like in Jupyter.
8. Logs at INFO and below are dropped by default, and adding `-l INFO` to my cmd-line doesn't seem to help. I was skeptical of my ability to show error logs to the user in every cell in which they might happen.

With all that said, `marimo` is my new favorite tool in the Python ecosystem. I find it far less frustrating than oldschool notebooks with the reactivity, autocomplete, mouse hover for docs, etc. The `uv` integration is friggin' awesome. On my Windows PC, it was less work for me to install uv, marimo, then pandas and plotly from your autosuggestions when i started coding in the notebook, than it was to figure out how to add `conda-forge` to my `Anaconda Navigator`. I love that the notebooks are git-able Python files.

I do wonder if the "variable df was defined by another cell" could be worked around by opting in to the latest cell being allowed to redefine vars. I try to use unique-but-short-names and often find myself accidentally referring to `df` again, after many cells of filtering it into various `xyzdf`'s. It wouldn't surprise me if this is a silent rejection point for a lot of more data-sciencey types. It seems clear to me that cell 40 saying "df = df.refine()" is topologically sortable to after the original definer; we have the cell's implicit order within the notebook as tiebreakers. I know it's not "pure" but UX is not always so mathematically elegant.

Best of luck with the road to v1 and thank you for the awesome tool and open planning!

**Reply by akshayka** [reply]
Hey @trenchant-dev — thanks so much taking the time to write such thoughtful and helpful feedback, and for the kind words.

1. We've gotten feedback on persisting UI element selections across runs a few times now. I see the value. Are you mostly interested in persisting the state of `ui.dataframe`? Or now that you know you need to copy the Python code, is that a satisfactory solution?

2. and 3. Happy to improve this. The next time you notice these discrepancies, if you could open a GitHub issue with a code example, that would be much appreciated.

4. Just merged a fix for `format_mapping`, thanks!

5. Thanks for the +1 — helpful to know that this made it hard for you to use marimo at work. We need to think through the UI for exposing per-notebook autorun, and how to distinguish it from the global configuration in the marimo editor.

6. Does `mo.output.append()` work for displaying multiple things?

7. We just added split cell to the right-click menu based on your feedback (#3534) 

8. We're probably doing something wonky with logs. Can look into that.

> I do wonder if the "variable df was defined by another cell" could be worked around by opting in to the latest cell being allowed to redefine vars.

I think you're right that this is the hardest thing to overcome for folks coming from Jupyter, and for some users it can be very hard. The main issue with letting cells redefine vars, using the cell's implicit order as a tiebreaker, is that rearranging cell order could change the DAG. To me this feels difficult to explain to users. It also leads to edge cases such as:

```
df # what is the value of df?
```

```
df = ...
```

```
df = ...
```


That said I would love to find a solution to this issue ...

**Reply by mscolnick** [reply]
8. We can look into making the default better. In the meantime, you need to set `logging.basicConfig`.

```python
import logging
level = logging.WARN
logging.basicConfig(level=level)
logger = logging.getLogger(__name__)
logger.setLevel(level)
```

[Demo notebook](https://marimo.app/l/38c6i6)
